design-documents/frontend/technical-vision.md

```diff
@@ -0,0 +1,140 @@
+
+1. First thing's first – what is the license?
+   - Is it open source?
+   - You may need to reach out to your legal department if things are a bit blurry.
+1. Does the package have a bright future?
+   - Popularity – are there a significant number of weekly downloads?
+   - Is the downloads line graph going up or down?
+1. Look for tests, ideally with 100% test coverage.
+1. Look at install size with [Package Phobia](https://packagephobia.now.sh/).
+1. Does it have good/clear documentation?
+1. Is it currently maintained?
+   - Look at the number of issues vs. pull requests.
+   - How long have these issues and PRs been open?
+   - When was the last comment made on them and by whom?
+   - **Reliability** – how serious are the issues?
+1. Search and compare this package with other similar packages to gauge which package is the best choice and feel free to reach out to other team members for guidance.
+
+### Resources
+- [spectrum-css](http://opensource.adobe.com/spectrum-css/2.6.0/docs/)
+
+## Styles
+
+Bundle optimization works best when CSS Modules are imported directly into the JavaScript components that use them (a better [TypeScript][] experience is being talked about in [this CRA issue](https://github.com/facebook/create-react-app/issues/5677)). The coupling here is a bit unfortunate for some projects (e.g., libraries), but pretty standard for applications. If there is a specific reason for ultimate **portability** in your project, consider the [container paradigm][] to decouple the styles from the components, each component defining its own styles via a styles interface.
+
+### Resources
+
+- [Adding CSS Modules to CRA](https://facebook.github.io/create-react-app/docs/adding-a-css-modules-stylesheet)
+- [css-modules-typescript-loader](https://www.npmjs.com/package/css-modules-typescript-loader)
+- [CSS Modules VSCode extension](https://marketplace.visualstudio.com/items?itemName=clinyong.vscode-css-modules)
+
+## Code Quality
+
+_Quality, Maintainability, Productivity._
+
+We should leverage the built-in code coverage reports that [Jest][] provides, but also the following GitHub integrations/apps:
+
+- Continuous integration with [Travis-CI](https://travis-ci.com/).
+- [CodeClimate](https://codeclimate.com/)
+  - [**Quality** by Code Climate](https://codeclimate.com/quality/) provides automated code review for test coverage, **maintainability** and more so that you can **save time** and merge with confidence.
+  - Code Climate’s engineering process insights and automated code review for GitHub and GitHub Enterprise help you ship **better software, faster**.
+
+### Resources
+- [Code Climate for TypeScript is here!](https://codeclimate.com/changelog/5a147fcac7d08102b700084e/)
+- [Code Climate | TSLint](https://docs.codeclimate.com/docs/tslint)
+
+## Performance Testing
+
+_TODO: Is there a way for us to determine if components overuse actions or API calls?_

```

Comment:
@kandy I reached out to you on Slack about this on Dec. 14th, but no response. Do you have any more details as to what your expectations are for performance testing a front-end application here? Personally, I think this is more of a DevOps concern, outside of what CodeClimate's tools already provide (mentioned above in the "Code Quality" section).

Reply:
I expect to see such hi-level elements here:
- All S0, S1 scenarios from end to end *MUST* be automated and should be monitored and compare metrics as total load time, average action time / step. Degradation of metrics *MUST* be prohibited.
- All common reusable component should be covered with separate isolated performance test.
- All measurement SHOULD be tested on average user data  (like Medium profile) and on edge case (like 10x) to see scalability gaps

Reply:
@kandy are you aware of any specific tools that would aid in meeting your requirements? I'm looking for more specific guidance here.

I agree that performance tests should be separate/isolated, but can you be specific about what libraries you would use for this? Also, when you say separate, do you mean a separate repo too?

If we can't lock down anything more specific would you be OK with me copying what you write above directly into this "Performance Testing" section and tagging you for more information?

Reply:
@kandy I have some more questions about your post:
1. What are S0, S1 scenarios? I'm no performance architect.
   - How do you automate them, exactly?
   - Wouldn't prohibiting degradation of metrics prevent almost any feature from being added? Doesn't seem realistic.
1. Define "common reusable component". Wouldn't that be 99% of the components, if written properly and decoupled?

Reply:
>  are you aware of any specific tools that would aid in meeting your requirements? I'm looking for more specific guidance here.

I am not a front-end specialist, so I can only offer sitespeed.io, but at a high level, even if the tool does not exist, it should be developed.

> I agree that performance tests should be separate/isolated, but can you be specific about what libraries you would use for this? 

Don't know why you ask me :)

> Also, when you say separate, do you mean a separate repo too?

I did not think about it and it does not relate directly to the performance. But when you asked me, I noticed that there was nothing in the document about managing dependencies between packages, but it should there

> What are S0, S1 scenarios? How do you automate them, exactly?

High severity scenario that MUST work in the product (for example in Magento it's "add product to cart and checkout flow"). 
We have Performance Acceptance Build that run for every PR, and compare execution time for the scenarios on base branch vs PR branch

> Wouldn't prohibiting degradation of metrics prevent almost any feature from being added? Doesn't seem realistic.

It prohibited without additional approve of Performance Architect and Product Owner.

> Define "common reusable component". Wouldn't that be 99% of the components, if written properly and decoupled?

I believe "will be used 3+ times" is good definition.

Reply:
I think after this conversation I can add a small paragraph that reads something like this:
> Ideally, for each front-end project, there will be a separate project that covers both performance and critical-path testing. This project is separate so that it doesn't cripple the short feedback loops of feature development. Additionally, we would do well to have additional performance testing around the most commonly used components (e.g., used in 3+ places).

Reply:
Addressed in 4f625a47b50e0c0ac67ccbf5fed03eff17b94d40